Add Magnetized FM Disk

carmaza opened the pull request:
## Proposed changes

Add magnetized Fishbone-Moncrief disk as initial data for GRMHD.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add magnetized Fishbone-Moncrief disk

carmaza committed:
Enable py modules to be called from AnalyticData/GrMhd

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+ * is the norm of the magnetic field in the fluid frame, with \f$v_i\f$ being
+ * the spatial velocity, and \f$W\f$ the Lorentz factor.
+ */
+class MagnetizedDisk {
```
MagnetizedFmDisk

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+ * A_\phi(r,\theta) \propto \text{max}(\rho(r,\theta) - \rho_\text{thresh}, 0),
+ * \f}
+ *
+ * where \f$\rho_\text{thresh}\f$ is an user-specified threshold density that
```
is a user-specified

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+    using type = double;
+    static constexpr OptionString help = {
+        "The radial coordinate of the inner edge of the disk."};
+  };
```
add a lower bound check of 0.0 to the radii

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+  struct ThresholdDensity {
+    using type = double;
+    static constexpr OptionString help = {
+        "Rest mass density below which matter is unmagnetized."};
```
Fractional rest...

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+  struct InversePlasmaBeta {
+    using type = double;
+    static constexpr OptionString help = {
+        "The max-magnetic-pressure-to-max-fluid-pressure ratio."};
```
The ratio of max magnetic pressure to max fluid pressure

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+                 InversePlasmaBeta>;
+
+  static constexpr OptionString help = {
+      "Magnetized fluid disk orbiting a Kerr black hole."};
```
Fishbone Moncrief disk

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+    return {tuples::get<Tags>(variables(x, tmpl::list<Tags>{}))...};
+  }
+
+  template <typename DataType>
```
this tag is in Hydro/Tags

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+  template <typename DataType>
+  auto variables(
+      const tnsr::I<DataType, 3>& x,
+      tmpl::list<hydro::Tags::SpecificEnthalpy<DataType>> /*meta*/) const
```
replace this with `Tag`

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+  auto variables(
+      const tnsr::I<DataType, 3>& x,
+      tmpl::list<hydro::Tags::SpecificEnthalpy<DataType>> /*meta*/) const
+      noexcept -> tuples::TaggedTuple<hydro::Tags::SpecificEnthalpy<DataType>>;
```
return type is `TaggedTuple<Tag>`

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+      tmpl::list<hydro::Tags::LorentzFactor<DataType>> /*meta*/) const noexcept
+      -> tuples::TaggedTuple<hydro::Tags::LorentzFactor<DataType>>;
+
+  template <typename DataType>
```
add template parameters `Tag, Requires<tmpl::list_contains_v<grmhd_tags<DataType>, Tag>> = nullptr`

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.hpp:
```diff
@@ -0,0 +1,271 @@
+      -> tuples::TaggedTuple<
+          hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>;
+
+  template <typename DataType>
```
keep this one, remove all the other specific ones

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.cpp:
```diff
@@ -0,0 +1,288 @@
+  auto magnetic_field =
+      make_with_value<tnsr::I<DataType, 3, Frame::NoFrame>>(x, 0.0);
+
+  const DataType rest_mass_density =
```
move to just before it is used

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.cpp:
```diff
@@ -0,0 +1,288 @@
+      r_squared += sqrt(square(r_squared) + a_squared * z_squared);
+      sin_theta_squared /= (r_squared + a_squared);
+
+      const double r = sqrt(r_squared);
```
r -> radius

kidder commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedDisk.cpp:
```diff
@@ -0,0 +1,288 @@
+      // B^i is proportional to derivatives of the magnetic potential.
+      // As done by SpEC and other groups, we approximate the derivatives
+      // with a 2nd-order centered finite difference expression.
+      const double small = 0.0001 * bh_mass_;
```
write out the formulas

kidder reviewed (changes requested):
group review with @nilsdeppe

nilsvu commented on docs/References.bib:
```diff
@@ -354,6 +354,24 @@ @article{Teukolsky2015ega
+  year           = "2016",
+  number         = "2",
+  pages          = "22",
+  doi            = "10.3847/0067-0049/225/2/22",
```
Add url field with this doi please (see the guidelines linked at the top of this file)

nilsvu commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/TestFunctions.py:
```diff
@@ -2,10 +2,13 @@
 # See LICENSE.txt for details.
 
 import numpy as np
+from AnalyticSolutions.RelativisticEuler import TestFunctions as re
```
I think better import the functions you need directly. We don’t have guidelines on this yet so this is only a suggestion, but I think it makes it easier to maintain this code.

nilsvu reviewed (commented):
Is this the initial data we want to evolve for the accretion disk? Probably want to label it `priority` then.

Not a review, only a few comments:

carmaza commented:
New commits:

- change name MagnetizedDisk -> MagnetizedFmDisk
- fixup after the first review
- add calculation of magnetic field normalization.

kidder commented:
I am happy with the fix-ups; @nilsdeppe @nilsleiffischer ?

nilsdeppe commented:
I don't know what order the commits are in and this needs a rebase.

carmaza commented:
The order is the order in which they were pushed. The normalization (last commit) was pushed in a new commit because it's code you haven't reviewed yet. In the end, I'll squash the last four commits.

carmaza committed:
Rename FM disk py module and remove fishbone prefix

carmaza commented:
The conflicts were due to the update in how we handle py modules (#1382). I added a separate commit updating the module of `FishboneMoncriefDisk` as in #1382, and I updated (and squashed right away) the names and includes of the module of `MagnetizedFmDisk`. The order of the commits is:

- Enable py modules to be called from AnalyticData/GrMhd
- Rename FM disk py module and remove fishbone prefix
- Add magnetized Fishbone-Moncrief disk
- fixup rename class
- fixup
- Add magnetic field normalization

The last commit is new code (which is why I didn't include in in the fixup commits) so feel free to ask for extra fix-ups. 

In the end, the four last commits should be squashed.

nilsvu commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.cpp:
```diff
@@ -0,0 +1,320 @@
+    const size_t normalization_grid_res) noexcept
+    : bh_mass_(bh_mass),
+      bh_spin_a_(bh_mass * bh_dimless_spin),
+      inner_edge_radius_(bh_mass * inner_edge_radius),
```
It‘s a bit confusing that the private variable includes a mass factor, but has the same name as the one in units of mass.

nilsvu commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.cpp:
```diff
@@ -0,0 +1,320 @@
+      background_spacetime_{
+          bh_mass_, {{0.0, 0.0, bh_dimless_spin}}, {{0.0, 0.0, 0.0}}},
+      kerr_schild_coords_{bh_mass, bh_dimless_spin} {
+  ASSERT(threshold_density_ > 0.0 and threshold_density_ < 1.0,
```
You should probably pass an `OptionContext` and raise `PARSE_ERROR`s (see e.g. CylindricalBlastWave)

wthrowe replied:
This ASSERT is for the case where this is not constructed via the option parser (and to catch bugs, of course) because the parser already enforces these limits.

nilsvu reviewed (commented):
I confirm the Python imports are handled correctly and I'm fine with squashing those changes. I also have a few more comments.

However, this is not a full review and I'd prefer not to be a reviewer on this PR. It would probably make more sense for someone who knows this analytic data better than me to review this.

kidder commented:
@carmaza your test is timing out in the debug builds. You can change the time for a timeout following the instructions at https://spectre-code.org/group__testingframeworkgroup under attributes.  If it works with a timeout smaller than 8, then adjust the timeout, otherwise, you will need to make the test faster.

carmaza commented:
@kidder Yes, I noticed. What actually takes long is to find the magnetic field for a large number of grid points (see the constructor in the cpp). If 8 doesn't work, I'll just reduce the number of grid points in the test.

carmaza commented:
Squashed and rebased on develop. I set the time for a timeout equal to 7.

Travis will complain because there is an include longer than 80 characters in the python module of the test. Let me know if you'd like me to modify `tools/FileTestDefs.sh` to address this.

nilsvu commented on tests/Unit/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.py:
```diff
@@ -0,0 +1,139 @@
+
+import numpy as np
+
+import PointwiseFunctions.AnalyticSolutions.RelativisticEuler.FishboneMoncriefDisk as fm_disk
```
You can always break the line with `\`s.

nilsdeppe replied:
To be concrete, from a brief test I did on my laptop you could do:
```python
import PointwiseFunctions.AnalyticSolutions.\
    RelativisticEuler.FishboneMoncrief as fm_disk
```

kidder commented:
@nilsleiffischer @nilsdeppe please look at

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.hpp:
```diff
@@ -0,0 +1,242 @@
+ * is the norm of the magnetic field in the fluid frame, with \f$v_i\f$ being
+ * the spatial velocity, and \f$W\f$ the Lorentz factor.
+ */
+class MagnetizedFmDisk {
```
A general design question:

Why did you choose not to inherit from the non-magnetized case and make it a member variable? It seems much of this is code duplication from the non-magnetized case. I haven't tested it so there might be something I'm missing, but it just seems like the amount of code here could be cut in half or more by using inheritance.

carmaza replied:
I think this decision was made last year when we were in a hurry for having something that works. I'd be happy to change the design if there is no urgency to run this this week :)

nilsdeppe replied:
Ah okay, I think it's worth trying at this point :) Here is what I think is necessary:
1. Inherit publicly of the FM disk class
2. Write the `variables` functions to forward to the base class functions
3. Write constructors, `operator==`, and a `pup` function that forward to the base class for the appropriate code.
4. Inherit the base class options. You may need to do something like `using Option = Base::Option;`, but I'm not sure.

I think this would remove close to 50% of the code and would be a great example of how to do it for other solutions :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.hpp:
```diff
@@ -0,0 +1,242 @@
+class MagnetizedFmDisk {
+ public:
+  using equation_of_state_type = EquationsOfState::PolytropicFluid<true>;
+  using background_spacetime_type = gr::Solutions::KerrSchild;
```
Is this needed for anything? I thought we got rid of it.

carmaza replied:
You're right! my bad.

carmaza committed:
Make FM disk variables and members protected

carmaza commented:
Last commit implements the inheritance. Eventually that commit should be squashed.

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/GrMhd/MagnetizedFmDisk.hpp:
```diff
@@ -164,74 +118,83 @@ class MagnetizedFmDisk {
+      noexcept {
+    // Can't store IntermediateVariables as member variable because we
+    // need to
+    // be threadsafe.
```
I think this can be moved onto the previous line during squashing :)

nilsdeppe reviewed (commented):
This looks great. I'm happy for you to squash :)

carmaza commented:
Squashed and rebased on develop.
